File: web/lib/mirror_resources.js
/* Static discovery only: no source script execution or network access. */
(function(root, factory) {
  if (typeof module === 'object' && module.exports) module.exports = factory();
  else root.MirrorResources = factory();
}(typeof self !== 'undefined' ? self : this, function() {
  const VERSION = '1.0', MAX_TEXT = 8 * 1024 * 1024, MAX_REFERENCES = 20000;
  function resolve(raw, base) {
    raw = String(raw || '').trim();
    if (!raw || raw.startsWith('#')) return null;
    try {
      const u = new URL(raw, base);
      if (!['http:', 'https:'].includes(u.protocol) || u.username || u.password) return null;
      u.hash = ''; return u.href;
    } catch (_) { return null; }
  }
  function entity(s) {
    return s.replace(/&(#x[\da-f]+|#\d+|amp|quot|apos|lt|gt);/gi, (m, x) => {
      if (x[0] !== '#') return ({amp:'&',quot:'"',apos:"'",lt:'<',gt:'>'})[x.toLowerCase()];
      const n = parseInt(x.slice(x[1].toLowerCase() === 'x' ? 2 : 1), x[1].toLowerCase() === 'x' ? 16 : 10);
      return n > 0 && n <= 0x10ffff ? String.fromCodePoint(n) : '\ufffd';
    });
  }
  function html(text) {
    const links = [], styles = []; let base = null, dynamic = false;
    // Tokenise tags with quoted > support and consume raw script bodies as one token.
    const token = /<!--[\s\S]*?(?:-->|$)|<(script|style)\b((?:"[^"]*"|'[^']*'|[^'">])*)>([\s\S]*?)(?:<\/\1\s*>|$)|<([a-z][\w:-]*)\b((?:"[^"]*"|'[^']*'|[^'">])*)>/gi;
    let m;
    while ((m = token.exec(text))) {
      const tag = (m[1] || m[4] || '').toLowerCase();
      if (!tag) continue;
      const a = {}; const attrs = m[2] || m[5] || '';
      for (const x of attrs.matchAll(/([^\s=/>]+)\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s>]+))/g)) {
        const key = x[1].toLowerCase(); if (!(key in a)) a[key] = entity(x[2] ?? x[3] ?? x[4]);
      }
      if (tag === 'base' && base === null && a.href) base = a.href;
      if (['a','area','link'].includes(tag) && a.href) links.push([a.href, tag === 'link' ? 'asset' : 'page']);
      if (['img','script','iframe','source','video','audio','input','embed','track'].includes(tag) && a.src) links.push([a.src, tag === 'iframe' ? 'page' : 'asset']);
      if (a.poster) links.push([a.poster, 'asset']);
      if (tag === 'object' && a.data) links.push([a.data, 'asset']);
      if (['img','source'].includes(tag) && a.srcset) {
        for (const x of a.srcset.matchAll(/(?:^|,)\s*(\S+)(?:\s+[^,]*)?/g)) links.push([x[1].replace(/,+$/, ''), 'asset']);
      }
      if (a.style) styles.push(a.style);
      if (tag === 'style') styles.push(m[3] || '');
      if (tag === 'script') dynamic = true;
    }
    return {links, styles, base, dynamic};
  }
  function discover(text, mime, url) {
    const result = {resources:[], unsupported:[]};
    if (text.length > MAX_TEXT) { result.unsupported.push('discovery_text_limit'); return result; }
    mime = mime.split(';')[0].trim().toLowerCase();
    if (!['text/html','application/xhtml+xml','text/css'].includes(mime)) return result;
    let links = [], base = url;
    if (mime !== 'text/css') {
      const parsed = html(text); base = resolve(parsed.base, url) || url; links = parsed.links;
      if (parsed.dynamic) result.unsupported.push('script_generated_content_not_evaluated');
    }
    const seen = new Set();
    for (const [raw, kind] of links) {
      const target = resolve(raw, base);
      if (target && !seen.has(target)) {
        if (seen.size >= MAX_REFERENCES) { result.unsupported.push('discovery_reference_limit'); break; }
        seen.add(target); result.resources.push({url:target, kind});
      } else if (!target && raw && !/^(#|data:)/.test(raw.trim())) result.unsupported.push('unsupported_reference_scheme');
    }
    result.unsupported = [...new Set(result.unsupported)].sort(); return result;
  }
  return {VERSION, discover, resolve};
}));

File: cli/mirror_resources.py
"""Static HTML/CSS discovery. No execution, fetching or third-party dependencies."""
from html.parser import HTMLParser
from urllib.parse import urljoin, urlsplit, urlunsplit
import re

VERSION = '1.0'
MAX_TEXT = 8 * 1024 * 1024
MAX_REFERENCES = 20000


def resolve(raw, base):
    raw = str(raw or '').strip()
    if not raw or raw.startswith('#'):
        return None
    try:
        url = urlsplit(urljoin(base, raw))
        if url.scheme not in ('http', 'https') or not url.hostname or url.username or url.password:
            return None
        return urlunsplit((url.scheme, url.netloc, url.path or '/', url.query, ''))
    except ValueError:
        return None


def srcset(value):
    candidates = []
    position = 0
    while position < len(value):
        while position < len(value) and (value[position].isspace() or value[position] == ','):
            position += 1
        start = position
        data_url = value[position:position + 5].lower() == 'data:'
        while position < len(value) and not value[position].isspace() and (data_url or value[position] != ','):
            position += 1
        if position > start:
            candidates.append(value[start:position].rstrip(','))
        while position < len(value) and value[position] != ',':
            position += 1
    return candidates


class _HTML(HTMLParser):
    def __init__(self):
        super().__init__(convert_charrefs=True)
        self.links = []
        self.base = None
        self.styles = []
        self.in_style = False
        self.dynamic = False

    def handle_starttag(self, tag, attrs):
        a = {}
        for key, value in attrs:
            a.setdefault(key, value)
        if tag == 'base' and self.base is None and a.get('href'):
            self.base = a['href']
        if tag in ('a', 'area', 'link') and a.get('href'):
            self.links.append((a['href'], 'page' if tag in ('a', 'area') else 'asset'))
        if tag in ('img', 'script', 'iframe', 'source', 'video', 'audio', 'input', 'embed', 'track') and a.get('src'):
            self.links.append((a['src'], 'page' if tag == 'iframe' else 'asset'))
        for attr in ('poster', 'data' if tag == 'object' else ''):
            if attr and a.get(attr):
                self.links.append((a[attr], 'asset'))
        if tag in ('img', 'source') and a.get('srcset'):
            self.links.extend((s, 'asset') for s in srcset(a['srcset']))
        if a.get('style'):
            self.styles.append(a['style'])
        if tag == 'style':
            self.in_style = True
        if tag == 'script':
            self.dynamic = True

    def handle_endtag(self, tag):
        if tag == 'style':
            self.in_style = False

    def handle_data(self, data):
        if self.in_style:
            self.styles.append(data)


def css_urls(text):
    """Scan comments, strings and URL/import tokens without treating strings as code."""
    string = r'"(?:\\.|[^"\\])*"|\'(?:\\.|[^\'\\])*\''
    tokens = re.compile(r'/\*[\s\S]*?(?:\*/|$)|url\(\s*(' + string + r'|(?:\\.|[^)\\])*)\s*\)|@import\s+(' + string + r')|' + string, re.I)
    out = []
    for m in tokens.finditer(text):
        raw = m.group(1) if m.group(1) is not None else m.group(2)
        if raw is None:
            continue
        raw = raw.strip()
        if raw[:1] in ('"', "'"):
            raw = raw[1:-1]
        def unescape(match):
            token = match.group(1)
            if re.fullmatch(r'[0-9a-fA-F]{1,6}\s?', token):
                number = int(token.strip(), 16)
                return chr(number) if 0 < number <= 0x10ffff else '\ufffd'
            return '' if token in ('\n', '\r') else token
        out.append(re.sub(r'\\([0-9a-fA-F]{1,6}\s?|[\s\S])', unescape, raw))
    return out


def discover(text, mime, url):
    """Return resolved references and explicit static-analysis limitations."""
    result = {'resources': [], 'unsupported': []}
    if len(text) > MAX_TEXT:
        result['unsupported'].append('discovery_text_limit')
        return result
    mime = mime.split(';')[0].strip().lower()
    if mime not in ('text/html', 'application/xhtml+xml', 'text/css'):
        return result
    links = []
    base = url
    if mime != 'text/css':
        parser = _HTML()
        parser.feed(text)
        base = resolve(parser.base, url) or url
        links = parser.links
        for style in parser.styles:
            links.extend((raw, 'asset') for raw in css_urls(style))
        if parser.dynamic:
            result['unsupported'].append('script_generated_content_not_evaluated')
    if mime == 'text/css':
        links = [(raw, 'asset') for raw in css_urls(text)]
    seen = set()
    for raw, kind in links:
        target = resolve(raw, base)
        if target and target not in seen:
            if len(seen) >= MAX_REFERENCES:
                result['unsupported'].append('discovery_reference_limit')
                break
            seen.add(target)
            result['resources'].append({'url': target, 'kind': kind})
        elif target is None and raw and not raw.strip().startswith(('#', 'data:')):
            result['unsupported'].append('unsupported_reference_scheme')
    result['unsupported'] = sorted(set(result['unsupported']))
    return result
